feat: implements uno cards

guilhermevrs opened the pull request:
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`yarn build`) was run locally and any changes were pushed
- [x] Lint (`yarn lint`) has passed locally and any fixes were made for failures


## Pull request type

<!-- Please do not submit updates to dependencies unless it fixes an issue. --> 

<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. --> 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 


## What is the current behavior?
N/A

Issue Number: Closes #110 


## What is the new behavior?
<!-- Please describe the behavior or changes that are being added by this PR. -->

- Components now have UnoCards
- Code accepts other cards

## Does this introduce a breaking change?

- [X] Yes
- [ ] No

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->

Cards now need a type discriminator

## Other information

N/A

## A picture tells a thousand words

<!-- Please include screenshots if relevant. -->
![image](https://user-images.githubusercontent.com/1570567/101405703-c68aab00-38d8-11eb-8c3f-0d2144a55a03.png)

guilhermevrs commented on packages/app/.eslintcache:
```diff
@@ -0,0 +1 @@
+[{"/home/guilhermevrs/git_clones/cards/packages/app/src/App.tsx":"1","/home/guilhermevrs/git_clones/cards/packages/app/src/components/pages/HomePage/HomePage.tsx":"2","/home/guilhermevrs/git_clones/cards/packages/app/src/index.tsx":"3"},{"size":414,"mtime":1606996749260,"results":"4","hashOfConfig":"5"},{"size":1631,"mtime":1607374988837,"results":"6","hashOfConfig":"5"},{"size":386,"mtime":16069 ...
```
I need to run gitignore for this one

tib-tib replied:
And also stop using `git add .` 😂 😛

guilhermevrs commented on packages/components/src/models/UnoCard.ts:
```diff
@@ -0,0 +1,32 @@
+export const ALL_UNO_COLORS = ['red', 'blue', 'green', 'yellow'] as const;
+export type UnoCardColor = typeof ALL_UNO_COLORS[number];
+
+export const ALL_UNO_VALUES = [
```
I find this much more elegant than using enums.
Let me know what you think, and if you are ok, I can refactor PlayingCard as well

guilhermevrs commented on packages/components/src/models/UnoCard.ts:
```diff
@@ -0,0 +1,32 @@
+export type UnoCardValue = typeof ALL_UNO_VALUES[number];
+
+const UNO_CARD_TYPE = 'UnoCard' as const;
+export interface UnoCard extends Card {
```
I consider ok to have a UnoCard like ``{value: "joker", color:"blue"}``. Even if the color doesn't change the visual, it will define the rest of the game anyhow.

tib-tib commented:
Oops, looks like you forgot to run the tests in the `app` package 😉  Probably we should set up husky properly to run tests of all packages before pushing

tib-tib commented:
![image](https://user-images.githubusercontent.com/10576609/101463279-827dc180-393d-11eb-9846-7175f5d8f8de.png)

It looks like the bottom of the card is cropped, could you check why?

tib-tib commented on packages/components/src/components/atoms/UnoCardFace/UnoCardFace.tsx:
```diff
@@ -0,0 +1,687 @@
+    plus2: () => renderPlus2(fill),
+    plus4: () => renderPlus4(),
+    joker: () => renderJoker(),
+  };
```
I don't really like this logic here, I'd rather have a dumb svg component like we have for playing cards. One reason is that we have it as a component because we couldn't figure out how to use and export directly an svg file, and if we want to try this again in the future we will be constrained by this logic.

guilhermevrs replied:
I think we should discuss about this topic - through here it won't be super effective.

I am not super fan of having three components to have the PlayingCard faces. I don't think the logic is complex enough to justify the split into three components.
Furthermore, I find a pitty that the SVG component from PlayingCard is obliged to render all possible `def`s to only then render a `use` tag that will reference one of them. The way I propose, I only render what's going to be used.

tib-tib replied:
Yeah I think I both agree and disagree with you 😄 
I think it would be better to discuss it in a call for sure, so let's try to have one this week 🙂

guilhermevrs replied:
As discussed, we will move the internal functions to sub-components

tib-tib commented on packages/components/src/components/organisms/Table/Table.tsx:
```diff
@@ -11,13 +11,21 @@ import { BaseCard } from '@atoms/BaseCard';
+import { isPlayingCard, PlayingCard } from '@models/PlayingCard';
+import { CardRenderer } from '@models/CardRenderer';
+import { Card } from '@models/Card';
+import { isUnoCard, UnoCard } from '@models/UnoCard';
```
shouldn't you see this warning in eslint as well? ⬇️

tib-tib commented on packages/components/src/store/slices/table/index.ts:
```diff
@@ -27,28 +27,34 @@ const initialState: TableState = {
 
 const defaultPosition: Position = { x: 0, y: 0 };
 
+const createCardState = (cards, dimensions, randomFace) => ({
```
Maybe name it `createInitialCardState`?

and add return type `CardState`

guilhermevrs commented:
- [ ] CardRenderer in a molecule
- [ ] CardStyle utils fn

tib-tib commented on packages/components/src/components/organisms/Table/Table.tsx:
```diff
@@ -61,43 +78,92 @@ export const Table: FunctionComponent<TableProps> = ({ height, width }) => {
+      } else if (isUnoCard(thisCard)) {
+        cardStyle = {
+          dimensions: UNO_CARD_DIMENSIONS,
+          borderRadius: 12,
```
From what I've seen `borderRadius: 16` would look better 😉

guilhermevrs commented:
I have normally addressed all the points we have discussed.
Plus, I have modified the eslint rules so that it ignores if you don't use arguments that starts with _

tib-tib commented on packages/components/src/models/UnoCard.ts:
```diff
@@ -0,0 +1,32 @@
+  'reverse',
+  'joker',
+] as const;
+export type UnoCardValue = typeof ALL_UNO_VALUES[number];
```
minor: youd didn't change those variables to pascal case in the end?

guilhermevrs replied:
I forgot that one :)

tib-tib commented on packages/components/src/utils/card-style.ts:
```diff
@@ -0,0 +1,51 @@
+ * @param tableDimentions the dimensions of the table
+ * @param customCardStyle a custom fn for style
+ */
+export const getCardStyleFn = (
```
Why puttin the suffix `Fn` to the function name? We already know it's a function 😄

guilhermevrs replied:
Changed to Factory

tib-tib commented on packages/components/src/utils/card-style.ts:
```diff
@@ -0,0 +1,51 @@
+  borderRadius?: number;
+}
+
+export type CustomCardSytleFn = (card: Card, currentStyle?: CardTypeStyle) => CardTypeStyle;
```
typo: `CustomCardSytleFn` => `CustomCardStyleFn`
Also, I would not use abbreviation in that case => `CustomCardStyleFunction`

guilhermevrs replied:
I changed to Factory
